src/Features/CSharp/Portable/Completion/CompletionProviders/SpeculativeTCompletionProvider.cs

```diff
@@ -58,46 +58,107 @@ private async Task<bool> ShouldShowSpeculativeTCompletionItemAsync(Document docu
+        {
+            var token = syntaxTree.FindTokenOnLeftOfPosition(position, cancellationToken);
+            var prevToken = token.GetPreviousTokenIfTouchingWord(position);
+
+            if (((allowAsyncKeyword || !prevToken.IsKindOrHasMatchingText(SyntaxKind.AsyncKeyword)) &&
+                syntaxTree.IsMemberDeclarationContext(position, contextOpt: null, validModifiers: SyntaxKindSet.AllMemberModifiers, validTypeDeclarations: SyntaxKindSet.ClassInterfaceStructTypeDeclarations, canBePartial: true, cancellationToken: cancellationToken)) ||
+                syntaxTree.IsStatementContext(position, token, cancellationToken) ||
+                syntaxTree.IsGlobalMemberDeclarationContext(position, SyntaxKindSet.AllGlobalMemberModifiers, cancellationToken) ||
+                syntaxTree.IsGlobalStatementContext(position, cancellationToken) ||
+                syntaxTree.IsDelegateReturnTypeContext(position, token, cancellationToken))
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        private static int WalkOutOfGenericType(SyntaxTree syntaxTree, int position, SemanticModel semanticModel, CancellationToken cancellationToken)
+        {
+            var spanStart = position;
+            var token = syntaxTree.FindTokenOnLeftOfPosition(position, cancellationToken);
+
+            if (syntaxTree.IsGenericTypeArgumentContext(position, token, cancellationToken, semanticModel))
             {
                 // Walk out until we find the start of the partial written generic
-                while (syntaxTree.IsInPartiallyWrittenGeneric(testPosition, cancellationToken, out var nameToken))
+                while (syntaxTree.IsInPartiallyWrittenGeneric(spanStart, cancellationToken, out var nameToken))
                 {
-                    testPosition = nameToken.SpanStart;
+                    spanStart = nameToken.SpanStart;
                 }
 
                 // If the user types Goo<T, automatic brace completion will insert the close brace
                 // and the generic won't be "partially written".
-                if (testPosition == position)
+                if (spanStart == position)
                 {
-                    var typeArgumentList = leftToken.GetAncestor<TypeArgumentListSyntax>();
-                    if (typeArgumentList != null)
-                    {
-                        if (typeArgumentList.LessThanToken != default && typeArgumentList.GreaterThanToken != default)
-                        {
-                            testPosition = typeArgumentList.LessThanToken.SpanStart;
-                        }
-                    }
+                    spanStart = token.GetAncestor<GenericNameSyntax>()?.SpanStart ?? spanStart;
+                }
+
+                var tokenLeftOfGenericName = syntaxTree.FindTokenOnLeftOfPosition(spanStart, cancellationToken);
```

Comment:
I admit I'm still not getting this bit -- if spanStart is the start of the generic name, why are we hopping further back?

Reply:
This is for qualified name. Say `A.B.C<...`, generic name is `C`, qualified name is `A.B`. Previously I added tests for `System.Func`, I'll add another test for `System.Collections.Generic.List<T>`

Reply:
Ah, this might be worth a comment in the code.

Reply:
Although I guess GitHub cut off the QualifiedName thing to the right below which made it a bit clearer.